src/java.base/share/classes/sun/security/ssl/SSLCipher.java

```diff
@@ -1859,10 +1859,23 @@ public Plaintext decrypt(byte contentType, ByteBuffer bb,
                 }
 
                 if (bb.remaining() <= tagSize) {
-                    throw new BadPaddingException(
-                        "Insufficient buffer remaining for AEAD cipher " +
+                    String msg = "Insufficient buffer remaining for AEAD cipher " +
                         "fragment (" + bb.remaining() + "). Needs to be " +
-                        "more than tag size (" + tagSize + ")");
+                        "more than tag size (" + tagSize + ")";
+
+                    // Check for unexpected plaintext alert.
+                    if (ContentType.ALERT.equals(ContentType.valueOf(contentType))) {
+                        msg = "Unexpected plaintext alert received. This can happen"
+                            + " during TLSv1.3 handshake if client doesn't receive"
+                            + " server_hello due to network timeout and tries to"
+                            + " close a connection by sending an alert message.";
+
+                        if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
+                            SSLLogger.info(msg);
+                        }
+                    }
+
+                    throw new BadPaddingException(msg);
```

Comment:
Is BadPaddingException still good enough for alert messages? Is the description accuracy for both client and server cases?

Reply:
Yes, I think `SSLProtocolException` would be more appropriate here. So we throw `GeneralSecurityException` in SSLCipher which is being caught by SSLEngineInputRecord/SSLSocketInputRecord and wrapped with SSLProtocolException. DTLSInputRecord on the other hand ignores GeneralSecurityException as it should (this issue doesn't apply to DTLS).